File: sdk/ts/src/vault.ts
// Vault API - daemon preferences endpoints
// list  - GET /preferences/list
// get   - GET /preferences/get?key=...
// set   - POST /preferences/set { key, value }
// delete - POST /preferences/delete { key }

import type { SisoulClient } from "./client.js";
import type {
  ListPreferencesResponse,
  Preference,
  VaultGetResponse,
} from "./types.js";

export class VaultAPI {
  constructor(private readonly client: SisoulClient) {}

  async list(): Promise<Preference[]> {
    const resp = await this.client.get<ListPreferencesResponse>("/preferences/list");
    return resp.items ?? [];
  }

  async get(key: string): Promise<string | null> {
    if (!key) throw new Error("vault.get: key required");
    const q = encodeURIComponent(key);
    const resp = await this.client.get<VaultGetResponse>(`/preferences/get?key=${q}`);
    return resp.value;
  }

  async set(key: string, value: string): Promise<void> {
    if (!key) throw new Error("vault.set: key required");
    await this.client.post<{ ok: boolean }>("/preferences/set", { key, value });
  }

  async delete(key: string): Promise<void> {
    if (!key) throw new Error("vault.delete: key required");
    await this.client.post<{ ok: boolean }>("/preferences/delete", { key });
  }

  /** 批量读 - daemon 没真 batch endpoint, 这里串行兜底, 后续 daemon 加 /preferences/multi-get 可换. */
  async multiGet(keys: string[]): Promise<Record<string, string | null>> {
    const out: Record<string, string | null> = {};
    for (const k of keys) {
      out[k] = await this.get(k);
    }
    return out;
  }
}

File: sdk/rust/src/vault.rs
//! Vault API.

use crate::client::SisoulClient;
use crate::errors::{Result, SisoulError};
use crate::types::{ListPreferencesResponse, Preference, VaultGetResponse};

pub struct VaultAPI<'a> {
    c: &'a SisoulClient,
}

impl<'a> VaultAPI<'a> {
    pub fn new(c: &'a SisoulClient) -> Self {
        Self { c }
    }

    pub fn list(&self) -> Result<Vec<Preference>> {
        let r: ListPreferencesResponse = self.c.get_json("/preferences/list", &[], false)?;
        Ok(r.items)
    }

    pub fn get(&self, key: &str) -> Result<Option<String>> {
        if key.is_empty() {
            return Err(SisoulError::InvalidArgument("vault.get: key required".into()));
        }
        let r: VaultGetResponse = self.c.get_json("/preferences/get", &[("key", key)], false)?;
        Ok(r.value)
    }

    pub fn set(&self, key: &str, value: &str) -> Result<()> {
        if key.is_empty() {
            return Err(SisoulError::InvalidArgument("vault.set: key required".into()));
        }
        let body = serde_json::json!({ "key": key, "value": value });
        self.c.post_unit("/preferences/set", &body, false)
    }

    pub fn delete(&self, key: &str) -> Result<()> {
        if key.is_empty() {
            return Err(SisoulError::InvalidArgument("vault.delete: key required".into()));
        }
        let body = serde_json::json!({ "key": key });
        self.c.post_unit("/preferences/delete", &body, false)
    }
}
